libseccomp-sys/build.rs

```diff
@@ -7,24 +7,203 @@ use std::env;
+        let lib_dir = build_bundled::build()?;
+        build_linked::link(Some(&lib_dir), Some("static"))
+    }
+    #[cfg(not(feature = "bundled"))]
+    {
+        // Otherwise, link with system library.
+        build_linked::link(None, None)
     }
+}
+
+#[cfg(feature = "bundled")]
+mod build_bundled {
+    use super::*;
+    use std::io::Write;
+    use std::path::{Path, PathBuf};
+
+    const LIBSECCOMP_VERSION: &str = "2.5.3";
+    const LIBSECCOMP_RELEASE_URL: &str = "https://github.com/seccomp/libseccomp/releases/download/vVERSION/libseccomp-VERSION.tar.gz";
+    const LIBSECCOMP_SHA256SUM: &str =
+        "59065c8733364725e9721ba48c3a99bbc52af921daf48df4b1e012fbc7b10a76";
 
-    let link_type = match env::var(LIBSECCOMP_LINK_TYPE) {
-        Ok(link_type) if link_type == "framework" => {
-            return Err("Seccomp is a Linux specific technology".into());
+    /// Downloads (if `LIBSECCOMP_SRC_PATH` is not set) and builds the libseccomp source.
+    pub fn build() -> Result<PathBuf, Box<dyn std::error::Error>> {
+        let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+        let lib_path = env::var(LIBSECCOMP_LIB_PATH);
+
+        // If lib path is set, skip build and use lib path directly.
+        if let Ok(path) = lib_path {
+            return Ok(PathBuf::from(path));
         }
-        Ok(link_type) => link_type, // static or dylib
-        Err(_) => String::from("dylib"),
-    };
 
-    println!("cargo:rustc-link-lib={}=seccomp", link_type);
+        // Determine whether to download or not:
+        // If src is set, use it. Otherwise, download and set src path.
+        let src_path = if let Ok(path) = env::var(LIBSECCOMP_SRC_PATH) {
+            PathBuf::from(path)
+        } else {
+            download_libseccomp(out_dir.as_path())?
+        };
+
+        // Create the directory that libseccomp's output will be written to.
+        // Technically, we don't have to do this and can extract it from `libseccomp/src/.libs/` after
+        // the build is done but it's cleaner this way.
+        //
+        // If the directory already exists, we delete it and start over. This should only
+        // happen if the build is being rerun because the environment variables changed, or the
+        // build itself failed.
+        let mut build_output_path = out_dir;
+        build_output_path.push("libseccomp-build-out");
+
+        if build_output_path.exists() {
+            std::fs::remove_dir_all(&build_output_path)?;
+        }
+        std::fs::create_dir(&build_output_path)?;
+
+        // Build source with configure, make, make install, setting the output
+        // directory to a directory we'll make in our package's out_dir.
+
+        // Get compiler flags info.
+        let cc_info = cc::Build::new().get_compiler();
+        let cc_env = cc_info.cc_env();
+        let cflags_env = cc_info.cflags_env();
+
+        // TODO: maybe could replace these commands with just using cc if the makefile isn't doing
```

Comment:
While looking into the gperf issue, I now agree that it would be hard to build libseccomp using only cc because libseccomp uses gperf to generate syscall tables. However, because it only uses it at compile time, perhaps we should just require the user to install it via their package manager; the point of this build script is to bundle everything that's necessary at runtime, not at build time.

That is, I think it's fine to require the user also have gperf in the same way they need to have curl and tar and sha256sum installed in their build environment, but not necessarily in their target environment. Similarly, we're supporting the user in cross-compiling, but we're not downloading the compiler for them.

We could add a simple check to see if it's installed and have a nice error message instead of waiting for the make command to fail.

Reply:
I tried uninstalling gperf and the error message seems good enough that we can probably leave it, actually.

```
  --- stderr
  configure: error: please install gperf
  Error: "Running `configure` failed while building libseccomp."
```